lib/rubygems/commands/update_command.rb

```diff
@@ -137,6 +136,9 @@ def fetch_remote_gems(spec) # :nodoc:
   def highest_installed_gems # :nodoc:
     hig = {} # highest installed gems
 
+    # Get only gem specifications installed as --user-install
+    Gem::Specification.dirs = Gem.user_dir if options[:user_install]
```

Comment:
I think this is the actual fix needed and none of the changes in `lib/rubygems/specification.rb` are actually needed?

Reply:
I don't know where the new methods are being used? :thinking: I see what you're trying to do, but I don't think you finished the implementation? You might need to loop only through the newly added `Specification.user_installed_stubs` in the line below this one.

Also, we would need a new test to make sure that default gems are not updated.

Reply:
Hmm 🤔 you are right, I mixed things up.

I'll move the changes of `specification.rb` to another/followup PR

Reply:
Yeah. This doesn't completely fix the bug, but it improves things over the current situation in my opinion, so let's do it. I'll open a separate ticket for the remaining issue.